### PR TITLE
203 conditional effects bmmfit

### LINKED ISSUE #203: [feature-request] Adapt S3 method `conditional_effects` to `bmmfit` objects.
### Discussed in https://github.com/venpopov/bmm/discussions/200

<div type='discussions-op-text'>

<sup>Originally posted by **GidonFrischkorn** April  4, 2024</sup>
The `conditional_effects` methods works ok for most `bmmodels`, but I think we could add some tweaks to integrate it more seamlessly into the workflow of `bmm`. For example, currently users need to know explicitly if a model parameter is specified as a `dpar` or `nlpar` when calling `conditional_effects`. Moreover, for `dpar` the `conditional_effects` are transformed back to the native scale using the specified links, this is not the case for `nlpars`.

Here are some examples, for `conditional_effects` plots for the two-parameter mixture model.

`conditional_effects(ZL_fit, effects = "setsize", dpar = "kappa1")` yields
image

whereas `conditional_effects(ZL_fit, effects = "setsize", nlpar = "kappa")` yields:
image

For `theta1` versus `thetat` i thought that having the plot on the probability scale is super useful, but again that only works when calling it via the `dpar`:  `conditional_effects(ZL_fit, effects = "setsize", dpar = "theta1")`
image

when called using the `nlpar`:  `conditional_effects(ZL_fit, effects = "setsize", nlpar = "thetat")`
image

Ideally, we have an implementation of `conditional_effects` for `bmmfit` objects that uses a `par` argument to specify for which model parameter the effects should be plotted and maybe add an additional `scale` argument to specify if the plots should be given on the `parameter`/`sampling` scale or the `native` scale. 

What do you think @venpopov?</div>

The goal is to have a `conditional_effects.bmmfit` method that eases the extractions of conditional effects from `bmmfit` objects.

### PR DESCRIPTION
**Pull request by GidonFrischkorn**
## Summary
- Implement `conditional_effects()` S3 method for `bmmfit` objects, providing an intuitive interface to visualize predictor effects on model parameters (#203)
- Add shared parameter helper functions (`.get_parameter_info()`, `.is_softmax_param()`, `link_transform()`) to `helpers-parameters.R`
- Update link specifications in mixture model objects to reflect actual transformations

### `conditional_effects.bmmfit()`
- Accepts model parameter names directly via `par` (e.g., `"kappa"`, `"thetat"`) and automatically routes to the correct brms `dpar`/`nlpar`
- Supports `scale = "native"` (default) to show parameters on their natural scale via inverse link transformations, or `scale = "sampling"` to show on the MCMC sampling scale
- Handles softmax-transformed parameters (`thetat`, `thetant` in `mixture3p`) correctly
- Filters internal model variables (e.g., fixed parameters, internal mixture components) from the output
- Works for single parameters or all estimated parameters at once (`par = NULL`)

### Link corrections in mixture model objects
- **`mixture2p`**: `thetat` link changed from `"identity"` to `"logit"` — the mixing weight is estimated on the logit scale, and the link specification now reflects this
- **`mixture3p`**: `thetat` and `thetant` links changed from `"identity"` to `"softmax"` — these weights go through a softmax transformation, which is now explicitly declared

These link corrections are necessary for `conditional_effects()` to apply the correct inverse transformations when displaying parameters on their natural scale.

### Shared helpers (`helpers-parameters.R`)
- `link_transform()`: forward/inverse link transformations for all brms link functions
- `.get_parameter_info()`: classifies a parameter as `dpar`/`nlpar`, retrieves its link, and detects softmax
- `.is_softmax_param()`: identifies parameters subject to softmax transformation

## Test plan
- [x] Unit tests for `link_transform()` round-trips across all supported links
- [x] Unit tests for `.is_softmax_param()` and `.get_parameter_info()`
- [x] Integration tests for `conditional_effects.bmmfit()` with SDM, mixture2p, and mixture3p models
- [x] Comprehensive tests covering scale transformations, softmax handling, multi-parameter mode, and edge cases

### CODE REVIEW

File: R/conditional_effects.R [hunk 1]
```diff
@@ -0,0 +1,580 @@
...
+#' @param x A bmmfit object (created by [bmm()])
+#' @param par Character string. Name of the model parameter to compute effects for.
+#'   This should be one of the parameter names from the original model specification
+#'   (see `names(x$bmm$model$parameters)`). If `NULL` (the default), conditional
+#'   effects are computed for all estimated (non-fixed) parameters.
+#' @param scale Character. Scale on which to show the parameter:
+#'   \describe{
+#'     \item{`"native"` (default)}{Show on natural scale using inverse link transformation.
+#'       For example, `kappa` with log link shown on exp scale, `thetat` with
+#'       identity link shown on probability scale (between 0 and 1).}
```
[the remark this answers is not in the document]
Comment (GidonFrischkorn) [reply R/conditional_effects.R]: Fixed. Updated to reference logit (mixture2p) and softmax (mixture3p) instead of identity

File: R/conditional_effects.R [hunk 2]
```diff
@@ -0,0 +1,580 @@
...
+      brms::posterior_linpred,
+      c(linpred_args, list(nlpar = "thetant"))
+    )
+
+    # softmax: exp(x_i) / (exp(thetat) + exp(thetant) + exp(0))
+    denom <- exp(draws_t) + exp(draws_nt) + 1
+    if (par == "thetat") {
+      softmax_draws <- exp(draws_t) / denom
+    } else {
+      softmax_draws <- exp(draws_nt) / denom
```
[the remark this answers is not in the document]
Comment (GidonFrischkorn) [reply R/conditional_effects.R]: Fixed. Now subtracts pmax(draws_t, draws_nt, 0) before exponentiating to prevent overflow.

File: R/conditional_effects.R [hunk 3]
```diff
@@ -0,0 +1,580 @@
...
+    }
+
+    summ <- .ce_summarize_draws(softmax_draws, prob = prob, robust = robust)
+    df$estimate__ <- summ$estimate
+    df$lower__ <- summ$lower
+    df$upper__ <- summ$upper
```
[the remark this answers is not in the document]
Comment (GidonFrischkorn) [reply R/conditional_effects.R:509]: Fixed. Added df$se__ <- summ$se after the other summary columns.

File: tests/testthat/test-conditional_effects.R [hunk 4]
```diff
@@ -0,0 +1,390 @@
+# Tests for conditional_effects.bmmfit() and its internal helpers
+# Tier 1: Unit tests (always run, no fitted model)
+# Tier 2: Fixture-based integration tests (skip if not interactive)
```
[the remark this answers is not in the document]
Comment (GidonFrischkorn) [reply tests/testthat/test-conditional_effects.R]: Fixed. Changed "skip if not interactive" to "skip on CRAN" to match actual skip conditions.

File: R/conditional_effects.R [hunk 5]
```diff
@@ -0,0 +1,580 @@
...
+  )
+  
+  model <- bmmfit$bmm$model
+  if (!is.null(model$other_vars$nt_features)) {
+    nt_features <- model$other_vars$nt_features
+    internal_patterns <- c(internal_patterns, paste0("^", gsub("([.|()\\^{}+$*?[]\\\\])", "\\\\\\1", nt_features), "$"))
+  }
+  if (!is.null(model$other_vars$nt_distances)) {
+    nt_distances <- model$other_vars$nt_distances
+    internal_patterns <- c(internal_patterns, paste0("^", gsub("([.|()\\^{}+$*?[]\\\\])", "\\\\\\1", nt_distances), "$"))
```
[the remark this answers is not in the document]
Comment (GidonFrischkorn) [reply R/conditional_effects.R]: Fixed. The character class [] was malformed. Corrected to ([][(){}^$*+?.|\\\\]) which properly places ] first in the class. Used base R gsub — no stringr dependency needed.

[the remark this answers is not in the document]
Comment (GidonFrischkorn) [reply R/conditional_effects.R]: Fixed. The character class [] was malformed. Corrected to ([][(){}^$*+?.|\\\\]) which properly places ] first in the class. Used base R gsub — no stringr dependency needed.

File: R/conditional_effects.R [hunk 6]
```diff
@@ -0,0 +1,580 @@
...
+#'   spaghetti = TRUE,
+#'   ndraws = 100
+#' )
+#' }
+conditional_effects.bmmfit <- function(x,
+                                       par = NULL,
+                                       scale = c("native", "parameter", "sampling"),
+                                       ...) {
+  stopif(!inherits(x, "bmmfit"), "x must be a bmmfit object")
+
```
[the remark this answers is not in the document]
Comment (GidonFrischkorn) [reply R/conditional_effects.R]: Fixed. Added x <- restructure(x) after the stopif check, consistent with summary.bmmfit and update.bmmfit.

**Comment by venpopov** [comment]
> ### Link corrections in mixture model objects
> * **`mixture2p`**: `thetat` link changed from `"identity"` to `"logit"` — the mixing weight is estimated on the logit scale, and the link specification now reflects this
> * **`mixture3p`**: `thetat` and `thetant` links changed from `"identity"` to `"softmax"` — these weights go through a softmax transformation, which is now explicitly declared
> 
> These link corrections are necessary for `conditional_effects()` to apply the correct inverse transformations when displaying parameters on their natural scale.

That's great, but can you confirm that this doesn't affect anything in an unexpected way? The reason I'm asking is because brms itself codes mixture parameters as "identity" despite the softmax link - it does this because in the underlying stan code it uses a logsumexp manual trick to ensure numerical stability. 

At this point I'm not sure which functions are aware of and use the "link" information from the bmmodel object. It might be that this is fine and it doesn't break anything, but I just want to make sure. I don't know if we have any tests that would catch an issue

File: tests/testthat/test-helpers-parameters.R
Comment (venpopov) [inline tests/testthat/test-helpers-parameters.R:1]: There is no need to use bmm::: to test functions

File: tests/testthat/test-conditional_effects.R [hunk 7]
```diff
@@ -0,0 +1,396 @@
+# Tests for conditional_effects.bmmfit() and its internal helpers
+# Tier 1: Unit tests (always run, no fitted model)
+# Tier 2: Fixture-based integration tests (skip on CRAN)
+# Tier 3: Model-fitting integration tests (skip on CRAN)
+
+library(testthat)
```
Comment (venpopov) [inline tests/testthat/test-conditional_effects.R]: don't load testthat, not needed

**Comment by GidonFrischkorn** [comment]
> > ### Link corrections in mixture model objects
> > 
> > * **`mixture2p`**: `thetat` link changed from `"identity"` to `"logit"` — the mixing weight is estimated on the logit scale, and the link specification now reflects this
> > * **`mixture3p`**: `thetat` and `thetant` links changed from `"identity"` to `"softmax"` — these weights go through a softmax transformation, which is now explicitly declared
> > 
> > These link corrections are necessary for `conditional_effects()` to apply the correct inverse transformations when displaying parameters on their natural scale.
> 
> That's great, but can you confirm that this doesn't affect anything in an unexpected way? The reason I'm asking is because brms itself codes mixture parameters as "identity" despite the softmax link - it does this because in the underlying stan code it uses a logsumexp manual trick to ensure numerical stability.
> 
> At this point I'm not sure which functions are aware of and use the "link" information from the bmmodel object. It might be that this is fine and it doesn't break anything, but I just want to make sure. I don't know if we have any tests that would catch an issue

Yes, I totally agree. I will check once more, but I remember we set up #189 exactly because for the models of the `1.0.0` release the `link` did not do anything with respect to the model estimation. So for `mixture2p`, `mixture3p`, `imm`, and the `sdm` the link only serves information purposes.

In the `m3` and the newly added evidence accumulation model, the `links` in the model object are used when estimating the model. Either by setting the link for the `brms::custom_family` or when converting the `bmmformula` to a `brmsformula`.

I still think we should advise users against tampering with the `links` as they are set for good reasons. The only exception is the `custom` version of the `m3`, because there users specify their own model parameters for which they should ideally provide reasonable links.

So far, the only issue I could see is, if we would specify custom initial values for the `mixture2p` or `mixture3p` model. In `helpers_inits.R` we use the links to specify initial values on the parameter scale, although the `bmmodel` object stores the initial values on the native scale. For now, this is no issue, as we do not specify initial values for the relevant models, but we might consider how to resolve this issue.

**Comment by venpopov** [comment]
good to know it's not an issue for now. But we should open an issue to figure out how to systematically treat links to make sure this doesn't become a problem down the road as we add more models and/or change or add other internal functionality. Easy to track for now but it will only become harder. This is independent of the current PR and is a future problem then.

Or rather than opening a new issue, generalize the #189

**Comment by GidonFrischkorn** [comment]
I now more comprehensively traced all the places model$links is read in the codebase. The short answer for `mixture2p` and `mixture3p` the change is safe.

In more detail:

The mixture models use `brms::mixture()`, not `brms::custom_family()`. The `model$links` field is never passed to `brms` for these models; `brms` handles the softmax internally in its Stan code. As said in my previous response, only DDM/cswald/ezdm pass links to `custom_family()`.

Currently there is also no effect on initial values. `helpers-inits.R` uses `model$links` inside `create_initfun()`, but only when `model$init_ranges` is non-NULL. Mixture models don't define `init_ranges`, so `create_initfun()` returns 1 early and the link field is never touched.

The links for the mixture models are only read by `summary.R` (cosmetic display), `helpers-parameters.R` (metadata), and now `conditional_effects.R` for back-transformation. These are all purely bmm-internal.

The one internal conflict is that `apply_links()` and `inv_link()` in `bmmformula.R` don't support "softmax", but they're only called for M3 models (check_formula.m3), never for mixture models.

**Comment by GidonFrischkorn** [comment]
> good to know it's not an issue for now. But we should open an issue to figure out how to systematically treat links to make sure this doesn't become a problem down the road as we add more models and/or change or add other internal functionality. Easy to track for now but it will only become harder. This is independent of the current PR and is a future problem then

Yes, I agree.

What do you think about updating the issue #189 accordingly.

**Comment by venpopov** [comment]
> > good to know it's not an issue for now. But we should open an issue to figure out how to systematically treat links to make sure this doesn't become a problem down the road as we add more models and/or change or add other internal functionality. Easy to track for now but it will only become harder. This is independent of the current PR and is a future problem then
> 
> Yes, I agree.
> 
> What do you think about updating the issue #189 accordingly.

yes, updating is better than opening a new issue - I had edited my comment to suggest that in the meantime :)

**Comment by GidonFrischkorn** [comment]
I also resolved the two issues you flagged.

I remember that I went back and forth with how to implement the tests without including a big model fit object to verify all functionality and in some of these explorations these seem to have been added accidentally.

File: R/conditional_effects.R [hunk 8]
```diff
@@ -0,0 +1,635 @@
...
+#'   fit,
+#'   par = "kappa",
+#'   effects = "setsize",
+#'   spaghetti = TRUE,
+#'   ndraws = 100
+#' )
+#' }
+conditional_effects.bmmfit <- function(x,
+                                       par = NULL,
+                                       scale = c("native", "parameter", "sampling"),
```
Comment (venpopov) [inline R/conditional_effects.R]: Is there a difference between "parameter" and "sampling"? If not, we should choose one.

Comment (GidonFrischkorn) [reply R/conditional_effects.R]: No.

I think we should opt for "sampling" as "parameter" is ambiguous. I remember that I struggled with labeling these options and then kept two, but deciding for one would be better.

Comment (venpopov) [reply R/conditional_effects.R]: Sampling is more obvious. I'll change it as I review the rest

**Comment by venpopov** [comment]
The integration test takes about 1 min to run which is way too long for an interactive sessions. I removed it - if you think its needed, it must use a precompiled stan model and sample very quickly.

File: R/conditional_effects.R [hunk 9]
```diff
@@ -0,0 +1,635 @@
...
+#'   effects = "setsize",
+#'   spaghetti = TRUE,
+#'   ndraws = 100
+#' )
+#' }
+conditional_effects.bmmfit <- function(x,
+                                       par = NULL,
+                                       scale = c("native", "sampling"),
+                                       ...) {
+  stopif(!inherits(x, "bmmfit"), "x must be a bmmfit object")
```
Comment (venpopov) [inline R/conditional_effects.R]: unnecessary check - S3 methods only run when the corresponding object is provided

**Review by venpopov — APPROVED** [review]
There are things that could be simplified, but it is fine as it is, so we can do that at some point in the future. It's a really nice addition to the package.

**Review by venpopov — APPROVED** [review]
There are things that could be simplified, but it's fine as it is, so we can do that at some point in the future. It's a really nice addition to the package
